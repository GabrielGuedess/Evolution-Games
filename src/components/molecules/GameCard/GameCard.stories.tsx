import { Story, Meta } from '@storybook/react/types-6-0';

import { GameCard, GameCardProps } from './GameCard';

export default {
  title: 'Molecules/GameCard',
  component: GameCard,
  args: {
    id: '1',
    name: 'God of War: Ragnar√∂k',
    genres: [{ name: 'Action-adventure ', id: '1', slug: 'action-adventure' }],
    developers: [
      { id: '1', name: 'Santa Monica Games', slug: 'santa-monica-games' },
    ],
    release_date: new Date('2022'),
    image:
      'https://i.pinimg.com/originals/34/f8/e5/34f8e55845212171e8bb35edc5b08018.jpg',
    price: 250.0,
    score: 5.0,
    platforms: [{ name: 'Plastation 5', slug: 'ps5', id: '1' }],
    slug: 'god-of-war-ragnarok',
    primary_color: '#9db4d0',
  },
} as Meta<GameCardProps>;

export const Default: Story<GameCardProps> = args => <GameCard {...args} />;
