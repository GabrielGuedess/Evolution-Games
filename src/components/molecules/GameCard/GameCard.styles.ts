import { rgba } from 'polished';
import media from 'styled-media-query';

import styled, { css } from 'styled-components';

type Color = {
  color: string;
};

export const Wrapper = styled.article`
  width: 100%;
  height: 22rem;
  position: relative;
  overflow: hidden;
  margin: 0 auto;
  border-radius: 0.4rem;

  ${media.greaterThan('medium')`
    width: 29.1rem;
    height: 100%;
    border-radius: 0;
  `}
`;

export const GameImage = styled.div`
  ${({ theme }) => css`
    width: 100%;
    min-height: 37.8rem;
    position: relative;
    cursor: pointer;
    filter: blur(0.5px);

    ${media.greaterThan('medium')`
      filter: none;
      background-color: ${theme.colors.gameDetails};
    `}
  `}
`;

export const CircleBlur = styled.div<Color>`
  ${({ color }) => css`
    display: none;
    position: absolute;
    top: -120%;
    left: 0;
    width: 100%;
    aspect-ratio: 1/1;
    border-radius: 50%;
    background-image: linear-gradient(180deg, ${color}, ${rgba(color, 0.2)});
    filter: blur(7rem);

    ${media.greaterThan('medium')`
      display: block;
    `}
  `}
`;

export const GameContent = styled.div`
  ${({ theme }) => css`
    width: 100%;
    height: 100%;
    display: flex;
    justify-content: space-between;
    flex-direction: column;
    padding: 2rem;
    position: absolute;
    gap: ${theme.spacings.xsmall};
    top: 0;
    background: linear-gradient(90deg, #010914 12.5%, rgba(4, 4, 12, 0.3) 100%);

    ${media.greaterThan('medium')`
      width: auto;
      min-height: 20.8rem;
      position: inherit;
      background-color: ${theme.colors.gameDetails};
      overflow: hidden;
    `}
  `}
`;

export const GameInfo = styled.div`
  width: 100%;
  display: flex;
  justify-content: space-between;
  cursor: pointer;
`;

export const Score = styled.div`
  margin-top: 1rem;
  position: relative;
`;

export const Percentage = styled.p`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.regular};
    font-style: normal;
    line-height: ${theme.spacings.small};
    color: ${theme.colors.whiteText};
    text-shadow: 0 0 1rem ${rgba(theme.colors.white, 0.5)};
    position: absolute;
    top: 0.9rem;
    left: 1.1rem;
  `}
`;

export const BoxHighlight = styled.div`
  max-width: calc(100% - 4rem);
`;

export const Title = styled.strong`
  ${({ theme }) => css`
    display: block;
    font-size: ${theme.font.sizes.xxmedium};
    line-height: ${theme.spacings.medium};
    font-weight: ${theme.font.regular};
    color: ${theme.colors.whiteText};
    font-style: normal;
    margin-bottom: ${theme.spacings.small};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.xbig};
    `}
  `}
`;

export const Genre = styled.p`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    line-height: ${theme.spacings.xsmall};
    font-weight: ${theme.font.regular};
    font-style: normal;
    color: ${theme.colors.secondary};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.medium};
    `}
  `}
`;

export const Developer = styled.p`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xxsmall};
    line-height: ${theme.spacings.xsmall};
    font-weight: ${theme.font.regular};
    font-style: normal;
    color: ${theme.colors.secondary};
    margin-top: 5px;

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.xsmall};
    `}
  `}
`;

export const GameFooter = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const Price = styled.p`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.medium};
    font-weight: ${theme.font.regular};
    line-height: ${theme.spacings.small};
    font-style: normal;
    color: ${theme.colors.whiteText};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.xxmedium};
    `}
  `}
`;

export const Platform = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 1rem;
`;
