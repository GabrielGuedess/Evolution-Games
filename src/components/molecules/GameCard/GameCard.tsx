/* eslint-disable react/no-unused-prop-types */
import Image from 'next/image';
import Link from 'next/link';

import {
  PlatformIcon,
  Platform,
} from 'components/atoms/PlatformIcon/PlatformIcon';

import theme from '../../../styles/theme';
import * as S from './GameCard.styles';

export type GameCardProps = {
  id?: string;
  title: string;
  slug: string;
  genre: string;
  developer: string;
  year: string;
  img: string;
  score: number;
  price: string;
  platform: Platform[];
  primaryColor: string;
};

export const GameCard = ({
  year,
  genre,
  title,
  price,
  img,
  score,
  slug,
  developer,
  platform,
  primaryColor,
}: GameCardProps) => {
  const { gameBad, gameAverage, gameGreat } = theme.colors;
  const formatValue = Math.trunc((score / 5) * 100);
  const dasharray = `${formatValue} ${100 - formatValue}`;
  const feedbackColors = [
    gameBad,
    gameBad,
    gameAverage,
    gameAverage,
    gameGreat,
    gameGreat,
  ];

  return (
    <S.Wrapper>
      <Link href={`/game/${slug}`} passHref>
        <S.GameImage>
          <Image
            src={img}
            alt={title}
            priority
            layout="fill"
            objectFit="cover"
          />
        </S.GameImage>
      </Link>
      <S.GameContent>
        <S.CircleBlur color={primaryColor} />

        <Link href={`/game/${slug}`} passHref>
          <S.GameInfo>
            <S.BoxHighlight>
              <S.Title>{title}</S.Title>
              <S.Genre>{genre}</S.Genre>
              <S.Developer>
                {developer}, {year}
              </S.Developer>
            </S.BoxHighlight>
            <S.Score>
              <svg
                viewBox="0 0 42 42"
                width={42}
                height={42}
                role="img"
                aria-label="Score"
              >
                <path
                  strokeDasharray={dasharray}
                  strokeDashoffset="-25"
                  pathLength="100"
                  d="M1 21a20 20 0 1 1 40 0 20 20 0 0 1-40 0"
                  fill="none"
                  stroke={feedbackColors[Math.trunc(score)]}
                  strokeWidth="2"
                />
              </svg>
              <S.Percentage>
                {score.toLocaleString('pt-BR', {
                  minimumFractionDigits: 1,
                  maximumFractionDigits: 1,
                })}
              </S.Percentage>
            </S.Score>
          </S.GameInfo>
        </Link>
        <S.GameFooter>
          <S.Price>{price}</S.Price>
          <S.Platform>
            {platform?.includes('playstation') && (
              <PlatformIcon platform="playstation" />
            )}

            {platform?.includes('xbox') && <PlatformIcon platform="xbox" />}

            {platform?.includes('pc') && <PlatformIcon platform="pc" />}
          </S.Platform>
        </S.GameFooter>
      </S.GameContent>
    </S.Wrapper>
  );
};
