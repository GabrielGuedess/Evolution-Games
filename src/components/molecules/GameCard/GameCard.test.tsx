import { renderWithTheme } from 'utils/tests/helpers';

import { screen } from '@testing-library/react';

import { GameCard, GameCardProps } from './GameCard';

const props: GameCardProps = {
  id: '1',
  name: 'God of War: RagnarÃ¶k',
  genres: ['Action-adventure'],
  developer: 'Santa Monica Games',
  releaseDate: new Date('11-09-2022'),
  image:
    'https://i.pinimg.com/originals/34/f8/e5/34f8e55845212171e8bb35edc5b08018.jpg',
  price: 250,
  score: 5.0,
  platform: 'ps4',
  primaryColor: '#f00',
  slug: '',
};

type Image = { src: string; alt: string };

jest.mock('next/image', () => ({
  __esModule: true,
  default: (image: Image) => <img src={image.src} alt={image.alt} />,
}));

describe('<GameCard />', () => {
  it('should render the heading', () => {
    // Arrange
    const { container } = renderWithTheme(<GameCard {...props} />);

    const title = screen.getByText(props.name);
    const genre = screen.getByText(props.genres.join(', '));
    const description = screen.getByText(
      `${props.developer}, ${new Date(props.releaseDate).getFullYear()}`,
    );
    const image = screen.getByRole('img', { name: props.name });

    // Assert
    expect(title).toBeInTheDocument();
    expect(genre).toBeInTheDocument();
    expect(description).toBeInTheDocument();
    expect(image).toHaveAttribute('src', props.image);

    expect(container.firstChild).toMatchSnapshot();
  });
});
