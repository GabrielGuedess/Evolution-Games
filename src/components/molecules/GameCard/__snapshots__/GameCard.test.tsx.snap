// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['<GameCard /> should render the heading 1'] = `
.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FFFFFF;
  -webkit-transition: color 0.3s linear;
  transition: color 0.3s linear;
}

.c14:hover {
  cursor: pointer;
  color: #7C3AED;
}

.c15 {
  height: 1.8rem;
}

.c0 {
  width: 100%;
  height: 22rem;
  position: relative;
  overflow: hidden;
  margin: 0 auto;
  border-radius: 0.4rem;
}

.c1 {
  width: 100%;
  min-height: 37.8rem;
  position: relative;
  cursor: pointer;
  -webkit-filter: blur(0.5px);
  filter: blur(0.5px);
}

.c3 {
  display: none;
  position: absolute;
  top: -120%;
  left: 0;
  width: 100%;
  aspect-ratio: 1/1;
  border-radius: 50%;
  background-image: linear-gradient(180deg,#f00,rgba(255,0,0,0.2));
  -webkit-filter: blur(7rem);
  filter: blur(7rem);
}

.c2 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 2rem;
  position: absolute;
  gap: 1.6rem;
  top: 0;
  background: linear-gradient(90deg,#010914 12.5%,rgba(4,4,12,0.3) 100%);
}

.c4 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  cursor: pointer;
}

.c9 {
  margin-top: 1rem;
  position: relative;
}

.c10 {
  font-size: 1.4rem;
  font-weight: 400;
  font-style: normal;
  line-height: 2.4rem;
  color: #F4F4F5;
  text-shadow: 0 0 1rem rgba(255,255,255,0.5);
  position: absolute;
  top: 0.9rem;
  left: 1.1rem;
}

.c5 {
  max-width: calc(100% - 4rem);
}

.c6 {
  display: block;
  font-size: 2.0rem;
  line-height: 3.2rem;
  font-weight: 400;
  color: #F4F4F5;
  font-style: normal;
  margin-bottom: 2.4rem;
}

.c7 {
  font-size: 1.4rem;
  line-height: 1.6rem;
  font-weight: 400;
  font-style: normal;
  color: #717B8A;
}

.c8 {
  font-size: 1rem;
  line-height: 1.6rem;
  font-weight: 400;
  font-style: normal;
  color: #717B8A;
  margin-top: 5px;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c12 {
  font-size: 1.6rem;
  font-weight: 400;
  line-height: 2.4rem;
  font-style: normal;
  color: #F4F4F5;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 1rem;
}

@media (min-width:768px) {
  .c0 {
    width: 29.1rem;
    height: 100%;
    border-radius: 0;
  }
}

@media (min-width:768px) {
  .c1 {
    -webkit-filter: none;
    filter: none;
    background-color: #010914;
  }
}

@media (min-width:768px) {
  .c3 {
    display: block;
  }
}

@media (min-width:768px) {
  .c2 {
    width: auto;
    min-height: 20.8rem;
    position: inherit;
    background-color: #010914;
    overflow: hidden;
  }
}

@media (min-width:768px) {
  .c6 {
    font-size: 2.5rem;
  }
}

@media (min-width:768px) {
  .c7 {
    font-size: 1.6rem;
  }
}

@media (min-width:768px) {
  .c8 {
    font-size: 1.2rem;
  }
}

@media (min-width:768px) {
  .c12 {
    font-size: 2.0rem;
  }
}

<article
  class="c0"
>
  <div
    class="c1"
    href="/game"
  >
    <img
      alt="God of War: Ragnarök"
      src="https://i.pinimg.com/originals/34/f8/e5/34f8e55845212171e8bb35edc5b08018.jpg"
    />
  </div>
  <div
    class="c2"
  >
    <div
      class="c3"
      color="#f00"
    />
    <div
      class="c4"
      href="/game"
    >
      <div
        class="c5"
      >
        <strong
          class="c6"
        >
          God of War: Ragnarök
        </strong>
        <p
          class="c7"
        >
          Action-adventure
        </p>
        <p
          class="0 c8"
        >
          Santa Monica Games
          , 
          2022
        </p>
      </div>
      <div
        class="c9"
      >
        <svg
          aria-label="Score"
          height="42"
          role="img"
          viewBox="0 0 42 42"
          width="42"
        >
          <path
            d="M1 21a20 20 0 1 1 40 0 20 20 0 0 1-40 0"
            fill="none"
            pathLength="100"
            stroke="#0FBF00"
            stroke-dasharray="100 0"
            stroke-dashoffset="-25"
            stroke-width="2"
          />
        </svg>
        <p
          class="c10"
        >
          5,0
        </p>
      </div>
    </div>
    <div
      class="1 c11"
    >
      <p
        class="2 c12"
      >
        R$ 250,00
      </p>
      <div
        class="3 c13"
      >
        <div
          class="c14"
          color="white"
        >
          <svg
            aria-label="PlayStation 4 Icon"
            class="c15"
            fill="none"
            role="figure"
            viewBox="0 0 96 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#a)"
              fill="currentColor"
            >
              <path
                d="M93.806 16.883h.187v-1.225h.444V15.5h-1.075v.16h.444v1.224ZM94.796 15.759l.373 1.124h.185l.373-1.124v1.124h.155v-1.384h-.226l-.388 1.182-.401-1.182h-.258v1.384h.187v-1.124Zm-30.78-2.19V7.85c0-1.167.544-2.002 1.706-2.002h7.167c.044 0 .086-.044.086-.087V4.622a.08.08 0 0 0-.006-.03h-8.764c-1.9.242-2.827 1.566-2.827 3.257v5.72c0 1.167-.56 2.002-1.706 2.002h-7.57c-.042 0-.085.03-.085.087v1.138a.1.1 0 0 0 .063.096h8.593c2.24-.06 3.342-1.489 3.342-3.323ZM34.57 16.891h2.517a.1.1 0 0 0 .063-.094V13.15c0-1.167.659-1.974 1.62-1.974h10.365c2.336 0 3.47-1.44 3.47-3.328 0-1.691-.928-3.015-2.828-3.256H34.524c-.005.01-.012.017-.012.029V5.76c0 .058.028.087.086.087H48.26c1.161 0 1.706.835 1.706 2.002 0 1.167-.545 2.003-1.706 2.003H37.45c-1.677 0-2.938 1.426-2.938 3.3v3.645c0 .046.02.083.058.095ZM72.933 15.572h13.763c.043 0 .071.028.071.057v1.168c0 .057.044.086.086.086h2.48c.045 0 .086-.029.086-.086v-1.139c0-.043.044-.086.087-.086h2.465c.057 0 .086-.044.086-.086v-1.153c0-.044-.029-.087-.086-.087h-2.466c-.042 0-.086-.043-.086-.086V5.847c0-.679-.251-1.11-.676-1.254h-.717c-.224.057-.467.16-.728.317l-14.823 9.149c-.574.36-.775.763-.631 1.066.114.26.443.447 1.09.447Zm2.15-1.528 11.483-7.102c.073-.058.216-.044.216.115v7.103c0 .043-.043.086-.086.086H75.155c-.1 0-.144-.028-.157-.057-.016-.058.012-.101.085-.145ZM9.615.001c1.119.209 2.218.506 3.312.817.625.181 1.25.37 1.87.57.973.31 1.943.66 2.825 1.184.395.24.775.511 1.099.842.351.345.64.749.85 1.193.385.803.516 1.7.555 2.582.019.576.018 1.156-.09 1.724-.09.494-.248.98-.512 1.41a2.553 2.553 0 0 1-.951.917c-.378.203-.81.3-1.238.292-.65.006-1.28-.2-1.863-.472-.01-1.937-.001-3.875-.004-5.812-.004-.35.02-.705-.051-1.05-.05-.269-.16-.539-.37-.72a1.073 1.073 0 0 0-.502-.236.514.514 0 0 0-.493.191c-.19.236-.24.55-.248.844-.001 5.187.002 10.374-.001 15.56L9.615 18.51V0ZM3.342 13.127c1.82-.644 3.637-1.295 5.458-1.936.003.714 0 1.428 0 2.143 0 .097.006.194-.006.29-1.439.51-2.875 1.03-4.313 1.542-.199.076-.403.161-.557.313-.072.071-.13.177-.09.279.048.12.171.186.285.236.315.128.66.146.996.161.374-.006.75-.034 1.117-.115a4.59 4.59 0 0 0 .723-.22c.614-.223 1.229-.447 1.844-.667.004.605 0 1.21.002 1.814-.002.12.005.242-.007.363-.755.135-1.52.227-2.288.24a13.663 13.663 0 0 1-4.565-.688c-.482-.14-.96-.325-1.363-.631-.221-.171-.42-.387-.52-.652a.899.899 0 0 1 .027-.7c.101-.216.278-.387.474-.52.45-.311.949-.55 1.453-.761.435-.185.886-.33 1.33-.491Z"
              />
              <path
                d="M17.983 12.313c.413-.047.828-.059 1.243-.074 1.365.006 2.737.175 4.043.585.256.078.504.178.76.26.454.166.91.367 1.274.694.219.2.408.469.41.775.013.25-.1.49-.262.674-.309.346-.716.584-1.126.791-.435.222-.904.366-1.36.535-2.79 1-5.58 2.002-8.368 3.004v-2.342c.003-.032-.014-.086.03-.094 2.123-.755 4.244-1.512 6.366-2.269.287-.107.596-.185.84-.38.1-.077.197-.202.148-.336-.051-.117-.173-.182-.285-.231-.318-.126-.663-.15-1.001-.16a5.008 5.008 0 0 0-1.566.239c-1.512.53-3.02 1.065-4.532 1.593v-2.515c1.094-.384 2.236-.617 3.386-.749Zm6.2 5.777a.959.959 0 0 1 .976 1.364.955.955 0 0 1-.87.544.95.95 0 0 1-.826-.477.965.965 0 0 1-.051-.851.968.968 0 0 1 .772-.58Zm-.071.177a.791.791 0 0 0-.598.613.796.796 0 1 0 .598-.613Z"
              />
              <path
                d="M23.883 18.491h.517c.097.001.205.024.274.099.069.078.075.192.056.29-.011.065-.068.103-.11.149.036.032.082.061.09.113.02.1 0 .204.032.301.02.035.04.07.042.11-.072.003-.143.001-.214.003-.048-.114-.012-.24-.043-.358-.011-.047-.06-.076-.107-.075-.117-.005-.234-.002-.35-.002-.002.145.004.29-.003.435a4.432 4.432 0 0 0-.183 0c-.003-.355 0-.71 0-1.065Zm.187.169v.294c.104 0 .208.002.313 0 .051 0 .111-.01.145-.053a.185.185 0 0 0 .003-.188c-.027-.044-.084-.049-.13-.052-.11-.003-.22-.001-.331-.001Z"
              />
            </g>
            <defs>
              <clippath
                id="a"
              >
                <path
                  d="M0 0h96v20H0z"
                  fill="currentColor"
                />
              </clippath>
            </defs>
          </svg>
        </div>
      </div>
    </div>
  </div>
</article>
`;
